RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,1033 @@
+
+#### Constraints
+
+Today, there are two relevant constraints in F# - `null` and `not struct`:
+
+```fsharp
+type C1<'T when 'T: null>() = class end
+type C2<'T when 'T: not struct>() = class end
+```
+
+Both are actually the same thing for interoperation purposes today, as they emit `class` as a constraint. See [Unresolved questions](nullable-reference-types.md#unresolved-questions) for more. C# 8.0 will interpret the `class` constraint as a non-nullable reference type, so we will have to change that in F#.
+
+If an F# class `B` has nullability as a constraint, then an inheriting class `D` also inherits this constraint, as per existing inheritance rules:
+
+```fsharp
+type B<'T when 'T: null>() = class end
+type D<'T when 'T: null>()
+    inherit B<'T>
+```
+
+That is, nullabulity constraints propagate via inheritance.
+
+If an unconstrained type parameter comes from an older C# assembly, it is assumed to be nullable, and warnings are given when non-nullability is assumed. If it comes from F# and it does not have the `null` constraint, then it is assumed to be non-nullable. If it does, then it is assumed to be nullable.
+
+Additionally, C# will introduce the `class?` constraint, which is sugar for an application of the `[<Nullable>]` attribute (see [Metadata](FS-1060-nullable-reference-types.md#net-metadata) for more).
+
+### A note on nullability and fundamental F# type representations
+
+Although the concept of nullability will surface as syntax and generate warnings, these annotations are ignored for all other purposes. For example:
+
+* Nullable annotations are ignored when deciding type equivalents, though warnings are emitted for mismatches.
+* Nullable annotations are ignored when deciding type subsumption, though warnings are emitted for mismatches.
+* Nullable annotations are ignored when deciding method overload resolution, though warnings are emitted for mismatches in argument and return types once an overload is committed.
+* Nullable annotations are ignored for abstract slot inference, though warnings are emitted for mismatches.
+* Nullable annotations are ignored when checking for duplicate methods.
+
+To re-iterate: nullable reference types are about separating distinguishing the implicit `null` from a reference type, but they are not a new _kind_ of reference type. They are still the same reference type and, despite warnings, can still compile when a nullability rule is violated.
+
+### Type inference
+
+Nullability should be propagated through type inference. Today, it will currently not be, so we will need to create a "nullness type variable" to represent the uncertainty, then unify it away as more information becomes available. These would be ignored for most purposes (just like nullability annotations) and would not be generalizable (i.e., committed to non-null in the absence of other information.
+
+Some examples:
+
+```fsharp
+// Inferred signature:
+//
+// val makeNullIfEmpty : str:string -> string | null
```

Comment:
```F#
let neverNull (str: string) =
    match str with
    | "" -> ""
    | _ -> makeNullIfEmpty(str)
```
How do I prevent a `string | null` inference (when user knows better than compiler)?
Is there some `!` operator as in C#?

Reply:
Yep, this would require non-nullability assertions. I'll make a note of that in this section.

To prevent this, you'd need to do something like this:

```fsharp
let neverNull (str: string) =
    match str with
    | "" -> ""
    | _ -> 
        makeNullIfEmpty str
         |> Unchecked.notNull
```

Where `Unchecked.notNull` (or whatever the mechanism is, perhaps `!`) will assert non-nullability.